File: backend/app/schemas/audit_log.py
"""Schemas for `/api/v1/audit-logs` (Wave 4-F).

Read-only projection — caller cannot mutate audit rows by design.
"""

from __future__ import annotations

from datetime import datetime
from typing import Any
from uuid import UUID

from pydantic import BaseModel, ConfigDict


class AuditLogRead(BaseModel):
    model_config = ConfigDict(from_attributes=True, extra="forbid")

    id: int
    actor_user_id: UUID | None = None
    role: str | None = None
    action: str
    target_table: str
    target_id: str
    method: str | None = None
    path: str | None = None
    status_code: int | None = None
    latency_ms: int | None = None
    ip_address: str | None = None
    user_agent: str | None = None
    request_body: dict[str, Any] | None = None
    before: dict[str, Any] | None = None
    after: dict[str, Any] | None = None
    created_at: datetime
    updated_at: datetime

File: frontend/lib/schemas/audit-log.ts
/**
 * Audit-log zod schemas — mirrors `backend/app/schemas/audit_log.py`.
 *
 * Only a read schema is exported; the audit table is server-managed.
 */
import { z } from 'zod';

export const auditLogReadSchema = z.object({
  id: z.number().int(),
  actor_user_id: z.string().uuid().nullable().optional(),
  role: z.string().nullable().optional(),
  action: z.string(),
  target_table: z.string(),
  target_id: z.string(),
  method: z.string().nullable().optional(),
  path: z.string().nullable().optional(),
  status_code: z.number().int().nullable().optional(),
  latency_ms: z.number().int().nullable().optional(),
  ip_address: z.string().nullable().optional(),
  user_agent: z.string().nullable().optional(),
  request_body: z.record(z.unknown()).nullable().optional(),
  before: z.record(z.unknown()).nullable().optional(),
  after: z.record(z.unknown()).nullable().optional(),
  created_at: z.string(),
  updated_at: z.string(),
});
export type AuditLogRead = z.infer<typeof auditLogReadSchema>;

export interface AuditLogPage {
  items: AuditLogRead[];
  total: number;
  limit: number;
  offset: number;
}
